tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/Logging/Serilog/Formatting/SerilogLogFormatter.cs

```diff
@@ -0,0 +1,216 @@
+                LogEventLevelDuck.Debug => DirectSubmissionLogLevelExtensions.Debug,
+                LogEventLevelDuck.Information => DirectSubmissionLogLevelExtensions.Information,
+                LogEventLevelDuck.Warning => DirectSubmissionLogLevelExtensions.Warning,
+                LogEventLevelDuck.Error => DirectSubmissionLogLevelExtensions.Error,
+                LogEventLevelDuck.Fatal => DirectSubmissionLogLevelExtensions.Fatal,
+                _ => throw new ArgumentOutOfRangeException(nameof(logLevel))
+            };
+#endif
+
+        private static LogPropertyRenderingDetails RenderProperties(JsonTextWriter writer, ILogEvent logEvent)
+        {
+            var haveSource = false;
+            var haveService = false;
+            var haveHost = false;
+            var haveTags = false;
+            var haveEnv = false;
+            var haveVersion = false;
+            foreach (var property in logEvent.Properties)
+            {
+                var duckProperty = property.DuckCast<KeyValuePairStringStruct>();
+                var name = duckProperty.Key;
+
+                haveSource |= LogFormatter.IsSourceProperty(name);
+                haveService |= LogFormatter.IsServiceProperty(name);
+                haveHost |= LogFormatter.IsHostProperty(name);
+                haveTags |= LogFormatter.IsTagsProperty(name);
+                haveEnv |= LogFormatter.IsEnvProperty(name);
+                haveVersion |= LogFormatter.IsVersionProperty(name);
+
+                LogFormatter.WritePropertyName(writer, name);
+                FormatLogEventPropertyValue(writer, duckProperty.Value);
+            }
+
+            var renderingDetails = new LogPropertyRenderingDetails(haveSource, haveService, haveHost, haveTags, haveEnv, haveVersion, messageTemplate: logEvent.MessageTemplate.Text);
+            return renderingDetails;
+        }
+
+        private static void FormatLogEventPropertyValue(JsonTextWriter writer, object value)
+        {
+            // format the value correctly depending on type
+            if (value.TryDuckCast<ScalarValueDuck>(out var scalar))
+            {
+                LogFormatter.WriteValue(writer, scalar.Value);
+                return;
+            }
+
+            if (value.TryDuckCast<SequenceValueDuck>(out var sequence))
+            {
+                FormatSequence(writer, sequence.Elements);
+                return;
+            }
+
+            if (value.TryDuckCast<StructureValueDuck>(out var structure))
+            {
+                FormatStructure(writer, structure.Properties, structure.TypeTag);
+                return;
+            }
+
+            if (value.TryDuckCast<DictionaryValueDuck>(out var dictionary))
+            {
+                FormatDictionary(writer, dictionary.Elements);
+                return;
+            }
+
+            throw new InvalidOperationException("Unknown value type " + value.GetType());
```

Comment:
Does this match library behavior? Is it better to try to succeed or do you think this merits an exception?

Reply:
It does: https://github.com/serilog/serilog/blob/13dd8cfc5ee47a841128beaef853b3798464f807/src/Serilog/Data/LogEventPropertyValueVisitor.cs#L47-L64

That said, we probably shouldn't throw here, and should just skip the unknown type (and log it).

Reply:
fixed in https://github.com/DataDog/dd-trace-dotnet/pull/1945/commits/a9a4bc438c34c4e63ad0fa1b8572a94aed49eda8